src/Compilers/CSharp/Test/Semantic/Semantics/PatternMatchingTests2.cs

```diff
@@ -2151,5 +2151,165 @@ public class C {
+            var model = compilation.GetSemanticModel(tree);
+            var patterns = tree.GetRoot().DescendantNodesAndSelf().OfType<PatternSyntax>().ToArray();
+            Assert.Equal(4, patterns.Length);
+
+            Assert.Equal("T2 (var t3, t4: T4 t4) { V5 : T6 t5 }", patterns[0].ToString());
+            var ti = model.GetTypeInfo(patterns[0]);
+            Assert.Equal("T1", ti.Type.ToTestDisplayString());
+            Assert.Equal("T2", ti.ConvertedType.ToTestDisplayString());
+
+            Assert.Equal("var t3", patterns[1].ToString());
+            ti = model.GetTypeInfo(patterns[1]);
+            Assert.Equal("T3", ti.Type.ToTestDisplayString());
+            Assert.Equal("T3", ti.ConvertedType.ToTestDisplayString());
+
+            Assert.Equal("T4 t4", patterns[2].ToString());
+            ti = model.GetTypeInfo(patterns[2]);
+            Assert.Equal("T4", ti.Type.ToTestDisplayString());
+            Assert.Equal("T4", ti.ConvertedType.ToTestDisplayString());
+
+            Assert.Equal("T6 t5", patterns[3].ToString());
+            ti = model.GetTypeInfo(patterns[3]);
+            Assert.Equal("T5", ti.Type.ToTestDisplayString());
+            Assert.Equal("T6", ti.ConvertedType.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PatternTypeInfo_02()
+        {
+            var source = @"
+public class C
+{
+    void M(object o)
+    {
+        if (o is Point(3, 4.0)) {}
+    }
+}
+class Point
+{
+    public void Deconstruct(out object o1, out System.IComparable o2) => throw null;
+}
+";
+            var compilation = CreateCompilation(source);
+            compilation.VerifyDiagnostics();
+            var tree = compilation.SyntaxTrees[0];
+            var model = compilation.GetSemanticModel(tree);
+            var patterns = tree.GetRoot().DescendantNodesAndSelf().OfType<PatternSyntax>().ToArray();
+            Assert.Equal(3, patterns.Length);
+
+            Assert.Equal("Point(3, 4.0)", patterns[0].ToString());
+            var ti = model.GetTypeInfo(patterns[0]);
+            Assert.Equal("System.Object", ti.Type.ToTestDisplayString());
+            Assert.Equal("Point", ti.ConvertedType.ToTestDisplayString());
+
+            Assert.Equal("3", patterns[1].ToString());
+            ti = model.GetTypeInfo(patterns[1]);
+            Assert.Equal("System.Object", ti.Type.ToTestDisplayString());
+            Assert.Equal("System.Int32", ti.ConvertedType.ToTestDisplayString());
+
+            Assert.Equal("4.0", patterns[2].ToString());
+            ti = model.GetTypeInfo(patterns[2]);
+            Assert.Equal("System.IComparable", ti.Type.ToTestDisplayString());
+            Assert.Equal("System.Double", ti.ConvertedType.ToTestDisplayString());
+        }
+
+        [Fact]
+        public void PatternTypeInfo_03()
+        {
+            var source = @"
+public class C
+{
+    void M(object o)
+    {
+        if (o is Point(3, 4.0) { Missing: Xyzzy }) {}
```

Comment:
>Point( [](start = 17, length = 6)

Unrelated to this PR, the current formatting rule is to put a space between type and positional sub-pattern. Please file an issue if you think it should be formatted like an invocation instead (ie. no space).

Reply:
https://github.com/dotnet/roslyn/issues/27268